docs/architecture/adr-064-abci-2.0.md

```diff
@@ -0,0 +1,447 @@
+
+Example:
+
+```go
+// VoteExtensionHandler implements an Oracle vote extension handler.
+type VoteExtensionHandler struct {
+	cdc   Codec
+	mk    MyKeeper
+	state VoteExtState // This could be a map or a DB connection object
+}
+
+// ExtendVoteHandler can do something with h.mk and possibly h.state to create
+// a vote extension, such as fetching a series of prices for supported assets.
+func (h VoteExtensionHandler) ExtendVoteHandler(ctx sdk.Context, req abci.RequestExtendVote) abci.ResponseExtendVote {
+	prices := GetPrices(ctx, h.mk.Assets())
+	bz, err := EncodePrices(h.cdc, prices)
+	if err != nil {
+		panic(fmt.Errorf("failed to encode prices for vote extension: %w", err))
+	}
+
+	// store our vote extension at the given height
+	//
+	// NOTE: Vote extensions can be overridden since we can timeout in a round.
+	SetPrices(h.state, req, bz)
+
+	return abci.ResponseExtendVote{VoteExtension: bz}
+}
+
+// VerifyVoteExtensionHandler can do something with h.state and req to verify
+// the req.VoteExtension field, such as ensuring the provided oracle prices are
+// within some valid range of our prices.
+func (h VoteExtensionHandler) VerifyVoteExtensionHandler(ctx sdk.Context, req abci.RequestVerifyVoteExtension) abci.ResponseVerifyVoteExtension {
+	prices, err := DecodePrices(h.cdc, req.VoteExtension)
+	if err != nil {
+		log("failed to decode vote extension", "err", err)
+		return abci.ResponseVerifyVoteExtension{Status: REJECT}
+	}
+
+	if err := ValidatePrices(h.state, req, prices); err != nil {
+		log("failed to validate vote extension", "prices", prices, "err", err)
+		return abci.ResponseVerifyVoteExtension{Status: REJECT}
+	}
+
+	// store updated vote extensions at the given height
+	//
+	// NOTE: Vote extensions can be overridden since we can timeout in a round.
+	SetPrices(h.state, req, req.VoteExtension)
+
+	return abci.ResponseVerifyVoteExtension{Status: ACCEPT}
+}
+```
+
+#### Vote Extension Propagation & Verification
+
+As mentioned previously, vote extensions for height `H` are only made available
+to the proposer at height `H+1` during `PrepareProposal`. However, in order to
+make vote extensions useful, all validators should have access to the agreed upon
+vote extensions at height `H` during `H+1`.
+
+Since CometBFT includes all the vote extension signatures in `RequestPrepareProposal`,
+we propose that the proposing validator manually "inject" the vote extensions
+along with their respective signatures via a special transaction, `VoteExtsTx`,
+into the block proposal during `PrepareProposal`. The `VoteExtsTx` will be
+populated with a single `ExtendedCommitInfo` object which is received directly
+from `RequestPrepareProposal`.
+
+For convention, the `VoteExtsTx` transaction should be the first transaction in
+the block proposal, although chains can implement their own preferences. For
+safety purposes, we also propose that the proposer itself verify all the vote
+extension signatures it receives in `RequestPrepareProposal`.
```

Comment:
Real apps shouldn't need to do this, since all those extensions are supposed to have been validated upon reception in `H-1`. Not that, at this point, all this info is coming from the _local_ CometBFT, not from the network.